File: Soal2.py
import numpy as np
# Metode A: Perhitungan secara eksak
S_loop = 0
for k in range(1, 21):
    S_loop += 1 / k

# Metode B: Setiap pembagian dibulatkan 5 angka desimal
nDigits = 5
terms = []

for k in range(1, 21):
    nilai = 1 / k
    nilai_bulat = round(nilai, nDigits)
    terms.append(nilai_bulat)

S_bulat = sum(terms)

# Metode C: Tanpa looping menggunakan fungsi sum
k = np.arange(1, 21)
S_vec = np.sum(1 / k)

# Nilai referensi
S_ref = S_vec

# Menghitung error
AE_loop = abs(S_ref - S_loop)
RE_loop = AE_loop / abs(S_ref)
RPE_loop = RE_loop * 100

AE_bulat = abs(S_ref - S_bulat)
RE_bulat = AE_bulat / abs(S_ref)
RPE_bulat = RE_bulat * 100

AE_vec = abs(S_ref - S_vec)
RE_vec = AE_vec / abs(S_ref)
RPE_vec = RE_vec * 100

print("Metode\t\t\tHasil\t\t\tAbsolute Error\tRelative Error\tRPE")
print(f"Eksak/Looping\t\t{S_loop:.15f}\t{AE_loop:.15f}\t{RE_loop:.10f}\t{RPE_loop:.10f}%")
print(f"Pembulatan\t\t{S_bulat:.15f}\t{AE_bulat:.15f}\t{RE_bulat:.10f}\t{RPE_bulat:.10f}%")
print(f"Tanpa Looping (sum)\t{S_vec:.15f}\t{AE_vec:.15f}\t{RE_vec:.10f}\t{RPE_vec:.10f}%")

File: Soal2.m
% Galat pada perhitungan 1/1 + 1/2 + 1/3 + ..... + 1/20
format long g

% a. Perhitungan secara eksak
S_loop = 0;

for k = 1:20
    S_loop = S_loop + 1/k;
end

% b. Masing-masing pembagian dibulatkan
nDigits = 5;
scale = 10^nDigits;

terms = 1 ./ (1:20);
terms = round(terms * scale) / scale;
S_bulat = sum(terms);

% c. Tanpa looping (menggunakan fungsi sum)
S_vec = sum(1 ./ (1:20));

% Nilai referensi
S_ref = S_vec;

% Menghitung error
AE_loop = abs(S_ref - S_loop);
RE_loop = AE_loop / abs(S_ref);
RPE_loop = RE_loop * 100;

AE_bulat = abs(S_ref - S_bulat);
RE_bulat = AE_bulat / abs(S_ref);
RPE_bulat = RE_bulat * 100;

AE_vec = abs(S_ref - S_vec);
RE_vec = AE_vec / abs(S_ref);
RPE_vec = RE_vec * 100;

fprintf('Metode\t\t\tHasil\t\t\tAbsolute Error\t\tRelative Error\t\tRPE\n');

fprintf('Eksak/Looping\t\t%.15f\t%.15f\t%.10f\t%.10f%%\n', ...
    S_loop, AE_loop, RE_loop, RPE_loop);

fprintf('Pembulatan\t\t%.15f\t%.15f\t%.10f\t%.10f%%\n', ...
    S_bulat, AE_bulat, RE_bulat, RPE_bulat);

fprintf('Tanpa Looping (sum)\t%.15f\t%.15f\t%.10f\t%.10f%%\n', ...
    S_vec, AE_vec, RE_vec, RPE_vec);
